// @uirouter/visualizer/lib/visualizer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var preact_1 = require("preact");
var StateSelector_1 = require("./selector/StateSelector");
exports.StateSelector = StateSelector_1.StateSelector;
var StateVisualizer_1 = require("./statevis/StateVisualizer");
exports.StateVisualizer = StateVisualizer_1.StateVisualizer;
var StateTree_1 = require("./statevis/tree/StateTree");
exports.StateTree = StateTree_1.StateTree;
var TransitionVisualizer_1 = require("./transition/TransitionVisualizer");
exports.TransitionVisualizer = TransitionVisualizer_1.TransitionVisualizer;
var visualizer = function (router) { return new Visualizer(router, {}); };
exports.visualizer = visualizer;
function unmountComponent(node) {
    var Nothing = function () { return null; };
    preact_1.render(preact_1.h(Nothing, null), document.body, node);
}
var DEFAULTS = {
    state: true,
    transition: true,
};
var Visualizer = /** @class */ (function () {
    function Visualizer(router, options) {
        this.router = router;
        this.name = "visualizer";
        options = Object.assign({}, DEFAULTS, options);
        if (options.state) {
            this.stateVisualizerEl = StateVisualizer_1.StateVisualizer.create(router);
        }
        if (options.transition) {
            this.transitionVisualizerEl = TransitionVisualizer_1.TransitionVisualizer.create(router);
        }
    }
    Visualizer.prototype.dispose = function (router) {
        this.stateVisualizerEl && unmountComponent(this.stateVisualizerEl);
        this.transitionVisualizerEl && unmountComponent(this.transitionVisualizerEl);
        this.stateVisualizerEl = null;
        this.transitionVisualizerEl = null;
    };
    return Visualizer;
}());
exports.Visualizer = Visualizer;
