// @uirouter/visualizer/lib-esm/transition/BreadcrumbArrow.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { h, Component } from "preact";
var BreadcrumbArrow = /** @class */ (function (_super) {
    __extends(BreadcrumbArrow, _super);
    function BreadcrumbArrow() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.handleClick = function () {
            return _this.props.toggleExpand();
        };
        return _this;
    }
    BreadcrumbArrow.prototype.iconClass = function () {
        var iconClasses = {
            running: 'uir-icon uir-spin uir-iconw-spinner',
            success: 'uir-icon uir-iconw-check',
            redirected: 'uir-icon uir-iconw-share',
            ignored: 'uir-icon uir-iconw-circle-o',
            error: 'uir-icon uir-iconw-close'
        };
        return iconClasses[this.props.status];
    };
    BreadcrumbArrow.prototype.render = function () {
        return !this.props.transition ? null : (h("div", { className: this.props.status + " uirTranVis_historyEntry", onClick: this.handleClick },
            h("div", { className: "uirTranVis_historyEntrySummary" },
                h("div", { className: "uirTranVis_transId" }, this.props.transition.$id),
                h("div", { className: "uirTranVis_status" },
                    this.props.status,
                    !this.props.message ? null : h("span", null,
                        ": ",
                        this.props.message)),
                h("div", { className: "uirTranVis_transName" },
                    h("i", { className: this.iconClass() }),
                    h("span", null, this.props.transition.to().name)))));
    };
    return BreadcrumbArrow;
}(Component));
export { BreadcrumbArrow };
